### PR TITLE
lottie: use internal append rect/ellipse func

### PR DESCRIPTION
**Pull request by mgrudzinska**
The lottie format defines the starting point for
drawing rectangles and ellipses differently than
the API available in the TVG. In the case of using dashes or trim paths, this leads to discrepancies. Therefore, when using the lottie builder, internal functions for appending rectangles and circles are utilized.

@issue: https://github.com/thorvg/thorvg/issues/1997

### CODE REVIEW

**Comment by hermet** [comment]
1. Is this a better approach that compromises renderer compatibility and then modifies the Lottie? Or can we simply tweak the SVG alone?
2. Have you reviewed the specifications for rectangles here? If not, we may need to update only the problematic circle.

**Comment by mgrudzinska** [comment]
> 1. Is this a better approach that compromises renderer compatibility and then modifies the Lottie? Or can we simply tweak the SVG alone?
> 2. Have you reviewed the specifications for rectangles here? If not, you may need to update only the problematic circle.

I'm not sure if I understand the question, but I'll try to answer. 
The renderer is set to a certain starting point for the circle and rectangle. In TVG, it is consistent with the SVG standard, though it could well be different. But as long as it's different from the Lottie standard, we need to do something about it so that Lottie files are displayed correctly.

Here, I proposed such a solution because it's very simple and we can easily change it in the future. In this way, we can also merge prs #1987 and #1996

As for the ultimate solution - I've read more about the Lottie standard and I think that the implementation of 'direction' will result in the solution proposed here being able to be modified as well.

And ofc I checked the lottie specs for rects. That's why I included some changes regarding rects into this pr as well.

@hermet if you think this isn't a good approach, please close this pr. I'm new in the lottie format, I'm probably missing many aspects of it right now.

**Comment by hermet** [comment]
> The renderer is set to a certain starting point for the circle and rectangle. In TVG, it is consistent with the SVG standard, though it could well be different.

@mgrudzinska The renderer itself is a open method, it could not follow any specification, I've never designed it from the get go, it is a just a pure graphics engine.

**Comment by hermet** [comment]
I'm considering the less change as possible for the best of both worlds.

File: src/loaders/lottie/tvgLottieBuilder.cpp [hunk 1]
```diff
@@ -371,9 +373,68 @@ static void _repeat(LottieGroup* parent, unique_ptr<Shape> path, RenderContext*
...
+            {x + w - rx + hrx, y}, {x + w, y + ry - hry}, {x + w, y + ry} //cubicTo
+    };
+
+    shape->appendPath(commands, commandsCnt, points, pointsCnt);
+}
+
+
+static void _appendRect(Shape* shape, float x, float y, float w, float h, float r)
+{
+    if (mathZero(r)) {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieBuilder.cpp:428]: suggestion: 2 lines are enough:
```
if (mathZero(r)) _appendSharpRect(shape, x, y, w, h, r);
else _appendRoundedRect(shape, x, y, w, h, r);
```

**Review by hermet — COMMENTED** [review]
thanks, just a few nit-picks.

File: src/loaders/lottie/tvgLottieBuilder.cpp [hunk 2]
```diff
@@ -371,9 +373,68 @@ static void _repeat(LottieGroup* parent, unique_ptr<Shape> path, RenderContext*
 }
 
 
+static void _appendSharpRect(Shape* shape, float x, float y, float w, float h, float r)
+{
+    constexpr int commandsCnt = 5;
+    PathCommand commands[commandsCnt] = {
+            PathCommand::MoveTo, PathCommand::LineTo, PathCommand::LineTo,
```
[the remark this answers is not in the document]
Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieBuilder.cpp]: sorry for that, changed everywhere

**Review by hermet — APPROVED** [review]

File: src/loaders/lottie/tvgLottieBuilder.cpp [hunk 3]
```diff
@@ -371,9 +373,65 @@ static void _repeat(LottieGroup* parent, unique_ptr<Shape> path, RenderContext*
...
+    constexpr int pointsCnt = 4;
+    Point points[pointsCnt] = {{x + w, y}, {x + w, y + h}, {x, y + h}, {x, y}};
+
+    shape->appendPath(commands, commandsCnt, points, pointsCnt);
+}
+
+static void _appendRoundedRect(Shape* shape, float x, float y, float w, float h, float r)
+{
+    constexpr int commandsCnt = 10;
+    PathCommand commands[commandsCnt] = {
```
Comment (hermet) [inline src/loaders/lottie/tvgLottieBuilder.cpp:393]: this command list style would be good to apply in tvgShape.cpp as well.

Comment (mgrudzinska) [reply src/loaders/lottie/tvgLottieBuilder.cpp:393]: @hermet  I'm not sure if I understand... in 'appendRect' we use functions from pImpl like 'moveTo', 'cubicTo' etc. would you like to replace them with 'append' from pImpl?

**Pull request merged**